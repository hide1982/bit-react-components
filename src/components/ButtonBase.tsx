import React, { forwardRef, useEffect, useRef } from "react";
import styled from "styled-components";

import useForkRef from "../utils/useForkRef";
import useRipple from "./useRipple";

type DefaultProps = Omit<
  React.DetailedHTMLProps<
    React.ButtonHTMLAttributes<HTMLButtonElement>,
    HTMLButtonElement
  >,
  "ref"
>;

interface StyleProps {
  disableFocusStyle?: boolean;
}

type Props = DefaultProps & StyleProps;

const StyledButton = styled.button<StyleProps>`
  position: relative;
  overflow: hidden;
  display: inline-flex;
  background-color: transparent;
  border: none;
  cursor: pointer;
  outline: none;
`;

const BaseButton = forwardRef<HTMLButtonElement, Props>((props, ref) => {
  const { disableFocusStyle, children, onClick, ...rest } = props;
  const { pulsate, setRippleProps, Ripple } = useRipple();

  const baseButtonRef = useRef<HTMLButtonElement>(null);
  const handleRef = useForkRef<HTMLButtonElement>(ref, baseButtonRef);

  useEffect(() => {
    if (baseButtonRef.current) {
      const elem = baseButtonRef.current;
      setRippleProps({
        top: elem.offsetTop,
        left: elem.offsetLeft,
        size: Math.max(elem.clientWidth, elem.clientHeight),
      });
    }
  }, [handleRef, baseButtonRef, setRippleProps]);

  const handleClick = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    if (onClick) onClick(e);
    pulsate(e);
  };

  return (
    <StyledButton
      ref={handleRef}
      disableFocusStyle={disableFocusStyle}
      onClick={handleClick}
      {...rest}
    >
      {children}
      <Ripple />
    </StyledButton>
  );
});

export default BaseButton;
