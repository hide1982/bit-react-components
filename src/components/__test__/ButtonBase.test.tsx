import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import styled from "styled-components";
import ButtonBase from "../ButtonBase";

describe("ButtonBase component", () => {
  test("snapshot", () => {
    render(<ButtonBase>button</ButtonBase>);
    expect(screen.getByTestId("button-base")).toMatchSnapshot();
  });

  test("Should Ripple invisible by default", () => {
    render(<ButtonBase>button</ButtonBase>);
    expect(screen.getByTestId("ripple")).toHaveStyle("display: none");
  });

  test("Should Ripple visible when the button is clicked", () => {
    render(<ButtonBase>button</ButtonBase>);
    fireEvent.click(screen.getByTestId("button-base"));
    expect(screen.getByTestId("ripple")).toHaveStyle("display: block");
  });

  test("Can set Disabled", () => {
    render(<ButtonBase disabled>button</ButtonBase>);
    expect(screen.getByTestId("button-base")).toBeDisabled();
  });

  test("Can use onClick event callback", () => {
    const onClick = jest.fn();
    render(<ButtonBase onClick={onClick}>button</ButtonBase>);
    fireEvent.click(screen.getByTestId("button-base"));
    expect(onClick).toBeCalledTimes(1);
  });

  test("Can change the Ripple styles", () => {
    render(
      <ButtonBase
        rippleColors={["red", "blue"]}
        rippleRadius="50%"
        rippleDuration="1s"
      >
        button
      </ButtonBase>
    );
    const Ripple = screen.getByTestId("ripple");
    expect(Ripple).toHaveStyle({
      borderColor: "red blue",
      borderRadius: "50%",
      animation: "ripple 1s linear",
    });
  });

  test("Can override styles by styled-component", () => {
    const StyledButtonBase = styled(ButtonBase)`
      background-color: yellow;
    `;
    render(<StyledButtonBase className="extra-class">button</StyledButtonBase>);
    expect(screen.getByTestId("button-base")).toHaveStyle(
      "background-color: yellow"
    );
  });

  test("Can append className", () => {
    render(<ButtonBase className="extra-class">button</ButtonBase>);
    expect(screen.getByTestId("button-base")).toHaveClass("extra-class");
  });

  test("Can use ref", () => {
    const ref = React.createRef<HTMLButtonElement>();
    render(<ButtonBase ref={ref}>button</ButtonBase>);

    expect(ref.current?.dataset?.testid).toBe("button-base");
  });
});
