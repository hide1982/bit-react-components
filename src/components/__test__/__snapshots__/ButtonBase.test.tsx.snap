// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonBase component snapshot 1`] = `
<button
  class="sc-gsTCUz hKTmWo"
  data-testid="button-base"
>
  button
  <span
    class="sc-bdfBwQ eKEvNE"
    data-testid="ripple"
  />
</button>
`;
